File: python/mapper.py
#!/usr/bin/env python3
# Interactive mapper for DS4 (Python)
import os
import time
try:
    import hid
except Exception:
    hid = None

print('Python mapper starting')
if hid is None:
    print('hid library not available; run in simulated mode by setting SIMULATE=1')

devices = hid.enumerate() if hid else []
print('HID devices found:', len(devices))

target = None
for d in devices:
    if 'Sony' in (d.get('manufacturer') or '') or 'Wireless Controller' in (d.get('product') or ''):
        target = d
        break

if not target:
    print('No DualShock-like device found; running simulated diffs.')

STATE = {'raw': None}

def print_diff(prev, cur):
    diffs = []
    for i in range(max(len(prev), len(cur))):
        a = prev[i] if i < len(prev) else 0
        b = cur[i] if i < len(cur) else 0
        if a != b:
            diffs.append({'idx':i, 'before':a, 'after':b})
    if diffs:
        print('Report diffs:', diffs)

if target:
    d = hid.device()
    d.open_path(target['path'])
    d.set_nonblocking(True)
    try:
        while True:
            r = d.read(64)
            if r:
                cur = list(r)
                if STATE['raw']:
                    print_diff(STATE['raw'], cur)
                STATE['raw'] = cur
            time.sleep(0.01)
    except KeyboardInterrupt:
        print('\nMapper stopped')
else:
    import random
    fake = [0]*16
    try:
        while True:
            idx = random.randrange(0, 15)
            mask = 1 << random.randrange(0,7)
            fake[idx] ^= mask
            if STATE['raw']:
                print_diff(STATE['raw'], fake)
            STATE['raw'] = fake.copy()
            time.sleep(0.7)
    except KeyboardInterrupt:
        print('\nMapper stopped')

File: server/mapper.js
#!/usr/bin/env node
// Interactive mapper for DualShock 4 (Node)
// Usage: MAP=1 node server/mapper.js

const HID = (() => {
  try {
    return require('node-hid');
  } catch (e) {
    return null;
  }
})();
const fs = require('fs');
const path = require('path');

const devices = HID ? HID.devices() : [];
console.log('HID devices:', devices.length);

const target = devices.find(
  (d) => /Sony/i.test(d.manufacturer || '') || /Wireless Controller/i.test(d.product || '')
);
if (!target) {
  console.log(
    'No DualShock-like device found. You can still run mapper in simulated mode with SIMULATE=1.'
  );
}

let device = null;
if (target) {
  try {
    device = new HID.HID(target.path);
    console.log('Opened device:', target.product || target.path);
  } catch (e) {
    console.error('Failed to open HID device:', e);
    device = null;
  }
}

const STATE = { raw: null };

function showHelp() {
  console.log('\nControls:');
  console.log('  Press a button on the controller — mapper prints byte diffs.');
  console.log('  Press Ctrl+C to stop.');
  console.log(
    '  To generate a mapping, press each button once while watching reported diffs and record the (byte,mask) pairs in a file called .ds4map.json'
  );
}

showHelp();

function printDiff(prev, cur) {
  const diffs = [];
  const len = Math.max(prev.length, cur.length);
  for (let i = 0; i < len; i++) {
    const a = prev[i] || 0;
    const b = cur[i] || 0;
    if (a !== b) diffs.push({ idx: i, before: a, after: b });
  }
  if (diffs.length) console.log('Report diffs:', JSON.stringify(diffs));
}

if (device) {
  device.on('data', (buf) => {
    const cur = Buffer.from(buf);
    if (STATE.raw) printDiff(STATE.raw, cur);
    STATE.raw = Array.from(cur);
  });
  device.on('error', (err) => console.error('Device error:', err));
} else {
  // simulation mode
  const simulate = () => {
    const fake = new Uint8Array(16);
    setInterval(() => {
      const idx = Math.floor(Math.random() * 15);
      const mask = 1 << Math.floor(Math.random() * 7);
      fake[idx] ^= mask;
      if (STATE.raw) printDiff(STATE.raw, fake);
      STATE.raw = Array.from(fake);
    }, 700);
  };
  simulate();
}

process.on('SIGINT', () => {
  console.log('\nMapper stopped.');
  process.exit(0);
});
